### PR TITLE
feat: add fork action and Monaco editor [DET-3568 DET-3127 DET-3660]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

partially depends on https://github.com/determined-ai/determined/pull/858

Screen Shot 2020-07-24 at 10 11 53 AM

## Commentary (optional)

with the current webpack set up App.test.tsx fails when it attempts to import Monaco editor. @hkang1 could we do without this app render test and keep these limited to unit tests? Our Cypress e2e tests cover already cover this (although slower)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: webui/react/config-overrides.js [hunk 1]
```diff
@@ -52,5 +54,26 @@ module.exports = override(
...
+      features: [
+        'suggest',
+        'quickOutline',
+        'wordHighlighter',
+        'colorDetector',
+        'codelens',
+        'parameterHints'
+      ],
+    })
+  ),
```
Comment (hkang1) [inline webui/react/config-overrides.js:72]: nice, didn't even have to change the override format!

File: webui/react/config-overrides.js [hunk 2]
```diff
@@ -52,5 +54,26 @@ module.exports = override(
...
 );
+
+module.exports = {
+  webpack: webpackConfig,
+  jest: (config, env) => ({...config, ...jestConfig}),
+  devServer: undefined,
```
Comment (hkang1) [inline webui/react/config-overrides.js:78]: **question:** what does setting `devServer` to `undefined` do? Is this needed for something specific?

File: webui/react/src/components/ExperimentActions.tsx [hunk 3]
```diff
@@ -85,7 +92,7 @@ const ExperimentActions: React.FC<Props> = ({ experiment, onSettled: updateFn }:
   };
 
   const actionButtons: ConditionalButton[] = [
-    { btn: <Button disabled key="fork" type="primary">Fork</Button> },
+    { btn: <Button key="fork" type="primary" onClick={onClick[Action.Fork]}>Fork</Button> },
```
Comment (hkang1) [inline webui/react/src/components/ExperimentActions.tsx:95]: **nit:** `button` instead of `btn`

File: webui/react/src/hooks/useRestApi.test.ts [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import * as io from 'io-ts';
 
-import { applyMappers } from './useRestApi';
+import { applyMappers } from 'utils/data';
```
Comment (hkang1) [inline webui/react/src/hooks/useRestApi.test.ts:3]: nice, good call

**Comment by hkang1** [comment]
This looks great, love the Monaco editor!

On styles, what about...:
- removing the `Using model definition from Experiment 7` line, since you can get back to the experiment just by hitting cancel or the overlay area.
- widening the modal box to avoid having to horizontally scroll
- if the 1st point is done, then maybe we can make the editor stretch to fill the entire modal body (basically no padding) but keep the padding for the error alert box

**Comment by hkang1** [comment]
the fork action was so fast, it looks like it just dismissed the modal but it actually took me to the forked experiment!

File: webui/react/src/utils/task.ts [hunk 5]
```diff
@@ -70,13 +70,15 @@ export const generateExperiments = (count = 10): ExperimentItem[] => {
...
+      };
       return {
         ...experimentTask,
-        config: {
-          description: experimentTask.name,
-          resources: {},
-          searcher: { metric: 'val_error', smallerIsBetter: true },
-        },
+        config,
+        configRaw: config,
```
Comment (hkang1) [inline webui/react/src/utils/task.ts:81]: 👍

**Comment by hkang1** [comment]
> with the current webpack set up App.test.tsx fails when it attempts to import Monaco editor. @hkang1 could we do without this app render test and keep these limited to unit tests? Our Cypress e2e tests cover already cover this (although slower)

@hamidzr agree with this, let's get rid of `App.test.tsx` and limit `test.tsx` to unit tests.

**Comment by hamidzr** [comment]
> > with the current webpack set up App.test.tsx fails when it attempts to import Monaco editor. @hkang1 could we do without this app render test and keep these limited to unit tests? Our Cypress e2e tests cover already cover this (although slower)
> 
> @hamidzr agree with this, let's get rid of `App.test.tsx` and limit `test.tsx` to unit tests.

I ended up looking more into react-app-rewired docs to add some jest changes to let us keep it. It seems to work fine, are they breaking for you?

**Comment by hamidzr** [comment]
> the fork action was so fast, it looks like it just dismissed the modal but it actually took me to the forked experiment!
> 
> **question:** are the logic for displaying the appropriate experiment actions not done yet? I noticed that the kill and cancel showed up for the forked experiment too when it hasn't activated yet.

react internal routing FTW.

re question: it should be. as soon as you fork the experiment goes into a paused state where you can kill or cancel. (existing behavior)

File: webui/react/src/hooks/useRestApi.test.ts (hunk 4 above)
Comment (hamidzr) [reply webui/react/src/hooks/useRestApi.test.ts:3]: the motivation was for the jest tests to not have to import useRestApi module to test. which was breaking when importing monaco

File: webui/react/config-overrides.js (hunk 2 above)
Comment (hamidzr) [reply webui/react/config-overrides.js:78]: nope just added it there to let us see it's available. could be a comment

File: webui/react/config-overrides.js (hunk 1 above)
Comment (hamidzr) [reply webui/react/config-overrides.js:72]: yeah their doc didn't show it clearly but the typescript definitions were clear you could do this

**Pull request merged**